### PR TITLE
Added Shopping cart feature. Possibility to add to the cart, update q…

### PR DESCRIPTION
**Pull request by Vadym-Ozi**
…uantity of added item, Delete cart item.

### CODE REVIEW

File: src/main/java/example/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+@RequiredArgsConstructor
+@Tag(name = "Shopping Cart Management", description = "Endpoints for managing shopping cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final CartItemService cartItemService;
+
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping
+    @Operation(summary = "Add item to cart")
+    public CartItemDto addItemToShoppingCart(@RequestBody @Valid RequestCartItemDto cartItemDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/ShoppingCartController.java:35]:
```suggestion
    public ShoppingRespondCartDto addItemToShoppingCart(Authentication authentication, @RequestBody @Valid RequestCartItemDto cartItemDto) {
```

File: src/main/java/example/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    @GetMapping
+    @Operation(summary = "Get shopping cart", description = "Get logged in user shopping cart")
+    public ShoppingRespondCartDto getCurrentUserShoppingCart() {
+        return shoppingCartService.getShoppingCart();
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @PutMapping("/items/{cartItemId}")
+    @Operation(summary = "Update shopping cart", description = "Update quantity of a book in the shopping cart")
+    public CartItemDto updateShoppingCart(@RequestBody @Valid RequestUpdateCartItemDto cartItemDto,
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/ShoppingCartController.java:49]:
```suggestion
    public ShoppingRespondCartDto updateShoppingCart(@RequestBody @Valid RequestUpdateCartItemDto cartItemDto,
```

File: src/main/java/example/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/ShoppingCartController.java:24]:
```suggestion
@RestController
@Validated
```
as you use Positive for Long id

File: src/main/java/example/dto/cartItem/RequestCartItemDto.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import jakarta.validation.constraints.DecimalMin;
+import jakarta.validation.constraints.Positive;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class RequestCartItemDto {
+    @Positive
```
Comment (Elena-Bruyako) [inline src/main/java/example/dto/cartItem/RequestCartItemDto.java:11]:
```suggestion
    @Positive
    @NotNull
```

File: src/main/java/example/model/ShoppingCart.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @JoinColumn(name = "user_id", nullable = false)
+    @OneToOne(fetch = FetchType.LAZY)
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL)
```
Comment (Elena-Bruyako) [inline src/main/java/example/model/ShoppingCart.java:35]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/example/model/ShoppingCart.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    private Set<CartItem> cartItems;
```
Comment (Elena-Bruyako) [inline src/main/java/example/model/ShoppingCart.java:36]:
```suggestion
    private Set<CartItem> cartItems = new HashSet<>();
```

File: src/main/java/example/model/ShoppingCart.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+import java.util.Set;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id = ?")
+@SQLRestriction("is_Deleted=false")
```
Comment (Elena-Bruyako) [inline src/main/java/example/model/ShoppingCart.java:24]:
```suggestion
@SQLRestriction("is_deleted=false")
```

File: src/main/java/example/service/impl/CartItemServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import example.service.CartItemService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
```
Comment (Elena-Bruyako) [inline src/main/java/example/service/impl/CartItemServiceImpl.java:24]: This service is redundant, all logic related to Shop cart only and should be in ShopCartService

File: src/main/java/example/controller/ShoppingCartController.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+@Tag(name = "Shopping Cart Management", description = "Endpoints for managing shopping cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final CartItemService cartItemService;
+
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping
+    @Operation(summary = "Add item to cart")
+    public CartItemDto addItemToShoppingCart(@RequestBody @Valid RequestCartItemDto cartItemDto) {
+        return cartItemService.addItemToCart(cartItemDto);
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/ShoppingCartController.java:36]:
```suggestion
        User user = (User) authentication.getPrincipal();
        return cartItemService.addItemToCart(cartItemDto, user.getId());
```

File: src/main/java/example/controller/ShoppingCartController.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping
+    @Operation(summary = "Get shopping cart", description = "Get logged in user shopping cart")
+    public ShoppingRespondCartDto getCurrentUserShoppingCart() {
+        return shoppingCartService.getShoppingCart();
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/ShoppingCartController.java:43]:
```suggestion
    public ShoppingRespondCartDto getCurrentUserShoppingCart(Authentication authentication) {
        User user = (User) authentication.getPrincipal();
        return shoppingCartService.getShoppingCart(user.getId());
```

File: src/main/java/example/controller/ShoppingCartController.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @PutMapping("/items/{cartItemId}")
+    @Operation(summary = "Update shopping cart", description = "Update quantity of a book in the shopping cart")
+    public CartItemDto updateShoppingCart(@RequestBody @Valid RequestUpdateCartItemDto cartItemDto,
+                                                   @Positive @PathVariable Long cartItemId) {
+        return cartItemService.updateCartItem(cartItemDto, cartItemId);
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/ShoppingCartController.java:51]:
```suggestion
        return cartItemService.updateCartItem(cartItemDto, cartItemId, user.getId());
```

File: src/main/java/example/controller/ShoppingCartController.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @DeleteMapping("/items/{cartItemId}")
+    @Operation(summary = "Delete item", description = "Delete book from shopping cart")
+    public void deleteItemFromShoppingCart(@PathVariable @Positive Long cartItemId) {
+        cartItemService.deleteCart(cartItemId);
```
Comment (Elena-Bruyako) [inline src/main/java/example/controller/ShoppingCartController.java:58]:
```suggestion
        cartItemService.deleteCart(cartItemId, user.getId());
```

File: src/main/resources/db/changelog/changes/10-create-cart-item-table.yaml [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+                  name: id
+                  type: BIGINT
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: shopping_cart_id
+                  type: BIGINT
+                  constraints:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/10-create-cart-item-table.yaml:19]: foreignKey?

File: src/main/resources/db/changelog/changes/10-create-cart-item-table.yaml [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+                    nullable: false
+              - column:
+                  name: book_id
+                  type: BIGINT
+                  constraints:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/10-create-cart-item-table.yaml:24]: foreignKey?

File: src/main/resources/db/changelog/changes/09-create-shopping-cart-table.yaml [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+                  name: id
+                  type: BIGINT
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: user_id
+                  type: BIGINT
+                  constraints:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/09-create-shopping-cart-table.yaml:19]: foreignKey?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/example/service/impl/ShoppingCartServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookService bookService;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final BookMapper bookMapper;
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/example/service/impl/ShoppingCartServiceImpl.java:34]: Add Transactional where is needed

File: src/main/java/example/service/impl/ShoppingCartServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingRespondCartDto updateCartItem(RequestUpdateCartItemDto cartItemDto,
+                                                 Long cartItemId,
+                                                 Long userId) {
+        CartItem cartItem = findCartItem(cartItemId, userId);
```
Comment (Elena-Bruyako) [inline src/main/java/example/service/impl/ShoppingCartServiceImpl.java:63]: Other option of implementation 
```
ShoppingCart cart = repository.findByUserId(userId);
CartItem cartItem = itemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())...
setQuantity
cartItemRepository.save(cartItem);
return shoppingCartMapper.toRespondDto(shoppingCart);
```

File: src/main/java/example/service/impl/ShoppingCartServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookService bookService;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final BookMapper bookMapper;
+
+    @Override
+    public CartItemDto addItemToCart(RequestCartItemDto cartItemDto, Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
```
Comment (Elena-Bruyako) [inline src/main/java/example/service/impl/ShoppingCartServiceImpl.java:36]: Other option of implementation
```
Book book = bookRepository.findById(itemDto.bookId())...
ShoppingCart cart = repository.findByUserId(userId);
cart.getCartItems().stream() -> check if book already exists - increase quantity, if no - set book to cartItem and set cartItem to shop cart 
shoppingCartRepository.save(cart);
return shoppingCartMapper.toDto(cart);
```

File: src/main/java/example/service/impl/ShoppingCartServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookService bookService;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final BookMapper bookMapper;
+
+    @Override
+    public CartItemDto addItemToCart(RequestCartItemDto cartItemDto, Long userId) {
```
Comment (Elena-Bruyako) [inline src/main/java/example/service/impl/ShoppingCartServiceImpl.java:35]:
```suggestion
    public ShoppingRespondCartDto addItemToCart(RequestCartItemDto cartItemDto, Long userId) {
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/example/service/impl/ShoppingCartServiceImpl.java (hunk 18 above)
Comment (Vadym-Ozi) [reply src/main/java/example/service/impl/ShoppingCartServiceImpl.java:36]: haven't thought that request already has a book id

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**